airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py

```diff
@@ -0,0 +1,257 @@
+        self._response_decoder = JsonDecoder()
+        self._error_message_parser = JsonErrorMessageParser()
+
+    # property moved from HttpStream
+    @property
+    def cache_filename(self) -> str:
+        """
+        Override if needed. Return the name of cache file
+        Note that if the environment variable REQUEST_CACHE_PATH is not set, the cache will be in-memory only.
+        """
+        return f"{self._stream_name}.sqlite"
+
+    # property moved from HttpStream
+    @property
+    def use_cache(self) -> bool:
+        """
+        Override if needed. If True, all records will be cached.
+        Note that if the environment variable REQUEST_CACHE_PATH is not set, the cache will be in-memory only.
+        """
+        return False
+
+    # public moved method from HttpStream
+    def request_session(self) -> requests.Session:
+        """
+        Session factory based on use_cache property and call rate limits (api_budget parameter)
+        :return: instance of request-based session
+        """
+        if self.use_cache:
+            cache_dir = os.getenv(ENV_REQUEST_CACHE_PATH)
+            # Use in-memory cache if cache_dir is not set
+            # This is a non-obvious interface, but it ensures we don't write sql files when running unit tests
+            if cache_dir:
+                sqlite_path = str(Path(cache_dir) / self.cache_filename)
+            else:
+                sqlite_path = "file::memory:?cache=shared"
+            return CachedLimiterSession(sqlite_path, backend="sqlite", api_budget=self._api_budget)  # type: ignore # there are no typeshed stubs for requests_cache
+        else:
+            return LimiterSession(api_budget=self._api_budget)
+
+    # public moved method from HttpStream
+    def clear_cache(self) -> None:
+        """
+        Clear cached requests for current session, can be called any time
+        """
+        if isinstance(self._session, CachedLimiterSession):
+            self._session.cache.clear()  # type: ignore # cache.clear is not typed
+
+    # private method moved from HttpStream
+    def _dedupe_query_params(self, url: str, params: Mapping[str, str]) -> Mapping[str, str]:
+        """
+        Remove query parameters from params mapping if they are already encoded in the URL.
+        :param url: URL with
+        :param params:
+        :return:
+        """
+        if params is None:
+            params = {}
+        query_string = urllib.parse.urlparse(url).query
+        query_dict = {k: v[0] for k, v in urllib.parse.parse_qs(query_string).items()}
+
+        duplicate_keys_with_same_value = {k for k in query_dict.keys() if str(params.get(k)) == str(query_dict[k])}
+        return {k: v for k, v in params.items() if k not in duplicate_keys_with_same_value}
+
+    # private method moved from HttpStream
+    def _create_prepared_request(
+        self,
+        http_method: str,
+        url: str,
+        dedupe_query_params: bool = True,
```

Comment:
It's feels like a small gotcha or catch that for the main entrypoint in `send_request()` the default for the `dedupe_query_params` parameter is False. But for this method it's default is true. We should make them consistent. I believe the existing `http.py` implementation always returns False so how you had it in `send_request()` was probably how we should represent it.

I'm really questioning why this even needs to be something we allow to be configured. The only place I see it overwritten is in low-code, but I feel like under the hood processing magic goes against the idea of declarative. @girarda am I crazy if we were to just 🪓 this config value?

Reply:
[The only reason we scoped the feature to low-code was we didn't have a good way to evaluate the blast radius](https://github.com/airbytehq/airbyte/issues/23497).

@brianjlai what should the behavior be if we 🪓 the parameter? I think we could just always deduplicate them, but we'll need to be more thorough with the regression testing to ensure it doesn't break any existing Python connectors

Reply:
Thanks for the catch. Will keep in for now, too.

Reply:
We can leave the param in since it is coupled to a requirement of low-code.

I do however, think we should make the default value `False`. That is how the HttpStream defines it since the base method in `http.py` is `must_deduplicate_query_params(self) = False`. And I also don't think always deduplicating params is a good default especially for Python sources.

In the Python CDK, if I have control of the inputs I'm passing when I invoke `send()` I should do my best to ensure correct inputs instead relying on the CDK to catch my mistakes or bugs.

Reply:
Noted. Updated default to False. Also fair point regarding putting onus on connector dev to make sure they have the correct inputs.